lighthouse-core/gather/gatherers/forms.js

```diff
@@ -0,0 +1,185 @@
+    }
+  }
+
+  return labelsArray;
+}
+
+/**
+ * @param {HTMLElement[]} formChildren
+ * @return { { inputs: { id: any; nodeName: any; name: any; placeholder: any; autocomplete: any; }[], labels: { id: any; nodeName: any; name: any; for: any; }[] } }
+ */
+/* istanbul ignore next */
+function getFormlessElements(formChildren) {
+  /** @type { { inputs: { id: any; nodeName: any; name: any; placeholder: any; autocomplete: any; }[], labels: { id: any; nodeName: any; name: any; for: any; }[] } } */
+  const formless = {
+    inputs: [],
+    labels: [],
+  };
+
+  for (const childElement of formChildren) {
+    if (childElement instanceof HTMLInputElement || childElement instanceof HTMLTextAreaElement
+      || childElement instanceof HTMLLabelElement || childElement instanceof HTMLSelectElement) {
+      if (childElement.form ) continue;
+    }
+
+    if (childElement instanceof HTMLInputElement || childElement instanceof HTMLTextAreaElement ) {
+      const inputAttributes = {
+        id: childElement.id,
+        nodeName: childElement.nodeName,
+        name: childElement.name,
+        placeholder: childElement.placeholder,
+        autocomplete: childElement.autocomplete,
+      };
+      formless.inputs.push(inputAttributes);
+    } else if (childElement instanceof HTMLSelectElement ) {
+      const selectAttributes = {
+        id: childElement.id,
+        nodeName: childElement.nodeName,
+        name: childElement.name,
+        placeholder: childElement.getAttribute('placeholder'),
+        autocomplete: childElement.autocomplete,
+      };
+      formless.inputs.push(selectAttributes);
+    } else if (childElement instanceof HTMLLabelElement) {
+      const labelAttributes = {
+        id: childElement.id,
+        nodeName: childElement.nodeName,
+        name: childElement.getAttribute('name'),
+        for: childElement.htmlFor,
+      };
+      formless.labels.push(labelAttributes);
+    }
+  }
+
+  return formless;
+}
+
+
+/**
+ * @return {LH.Artifacts['Forms']}
+ */
+/* istanbul ignore next */
+function collectFormElements() {
+  // @ts-ignore - put into scope via stringification
+  const formElements = getElementsInDocument('form'); // eslint-disable-line no-undef
+  // @ts-ignore - put into scope via stringification
+  const inputs = getElementsInDocument('input'); // eslint-disable-line no-undef
+  // @ts-ignore - put into scope via stringification
+  const labels = getElementsInDocument('labels'); // eslint-disable-line no-undef
+  // @ts-ignore - put into scope via stringification
+  const textareas = getElementsInDocument('textarea'); // eslint-disable-line no-undef
+  // @ts-ignore - put into scope via stringification
+  const selects = getElementsInDocument('select'); // eslint-disable-line no-undef
+
+  const formChildren = inputs.concat(labels, textareas, selects);
+  const formless = getFormlessElements(formChildren);
+
+  const forms = formElements.map(/** @param {HTMLFormElement} formElement */ (formElement) => {
```

Comment:
there's a lot of duplication w.r.t. getting the inputs and labels for `form` and `formless` sets of elements. I think all off this code can be shared if the logic focused on the input elements first, like:

```
const forms = new Map();
const formless = {inputs: [], labels: []};

const els = getElementsInDocument('label textarea select');

for every el:
  form = forms.get(el.form)
  if !form then create one and add to map
  
  based on what type of element `el` is, add to form.inputs or labels

return [...forms.values()];
```